src/Microsoft.AspNet.Mvc.TagHelpers/LinkTagHelper.cs

```diff
@@ -205,10 +205,20 @@ public class LinkTagHelper : TagHelper
         /// <inheritdoc />
         public override void Process(TagHelperContext context, TagHelperOutput output)
         {
+            string resolvedUrl;
+
             // Pass through attribute that is also a well-known HTML attribute.
             if (Href != null)
             {
                 output.CopyHtmlAttribute(HrefAttributeName, context);
+
+                // Resolve any application relative URLs (~/) now so they can be used in comparisons later.
+                if (TryResolveUrl(Href, tryEncodeApplicationPath: false, resolvedUrl: out resolvedUrl))
+                {
+                    Href = resolvedUrl;
+                }
+
+                ProcessUrlAttribute(HrefAttributeName, output);
```

Comment:
In addition the `AppendVersion` handling expects the `href` attribute to be either `null` or a `string`. Should handle the `HtmlString` case similar to `ProcessUrlAttribute()`. Existing issue but worth fixing while you're messing with this.